lib/Sema/CSApply.cpp

```diff
@@ -4275,8 +4275,15 @@ namespace {
     bool hasForcedOptionalResult(ExplicitCastExpr *expr) {
       const auto *const TR = expr->getCastTypeRepr();
       if (TR && TR->getKind() == TypeReprKind::ImplicitlyUnwrappedOptional) {
-        auto *locator = cs.getConstraintLocator(
-            expr, ConstraintLocator::ImplicitlyUnwrappedDisjunctionChoice);
+        ConstraintLocator *locator;
+        if (isExpr<CoerceExpr>(expr)) {
+          locator = cs.getConstraintLocator(
+              expr, {LocatorPathElt::CoercionOperand(),
```

Comment:
Do we actually need `CoercionOperand` here? Isn't the result of coercion being considered here instead?

Reply:
Yeah, otherwise it asserts because in `getDisjunctionChoice` for because that is the locator path expected for the disjunction. I thought it was expected since we  both call `addDisjunctionConstraint` and `createTypeVariableAndDisjunctionForIUOCoercion` at that locator in `CSGen`.

Reply:
So maybe we should just use `CoerceExpr` directly for call to `createTypeVariableAndDisjunctionForIUOCoercion`? It seems more reasonable considering that we are unwrapping the cast type.

Reply:
Wouldn't that require to have handle both(CoercionOperand and directlyAt<CoerceExpr>) on diagnostics? Let me try that and see

Reply:
It seems that there was no problem, but I'm relying on that to be covered by the tests. If that is ok, I can commit the changes =]

Reply:
We can always fix it if some cases come up in the future :)

Reply:
Fixed!